### ISSUE TITLE
sql: TTL does not appear to be ... TTLing in 22.1 alpha.3

### ISSUE THREAD

**Issue by rmloveland** [cockroachdb/cockroach#77843]
tl;dr: set table TTL to 3 minutes.  Insert some rows.  Wait more than 3 minutes, and `SELECT *`.  Rows not deleted after 3 minutes.

System info:

```
$ uname -a
Darwin crlMBP-C02F51KPML7JMTI2.local 21.3.0 Darwin Kernel Version 21.3.0: Wed Jan  5 21:37:58 PST 2022; root:xnu-8019.80.24~20/RELEASE_X86_64 x86_64
```

Start `cockroach demo`:

```
$ cockroach demo --insecure --global --nodes 9
... snip ...
# Server version: CockroachDB CCL v22.1.0-alpha.3 (x86_64-apple-darwin19, built 2022/03/14 15:31:44, go1.17.6) (same version as client)
# Cluster ID: 99012dc8-a723-4ef8-afea-e57f71f61bee
# Organization: Cockroach Demo
```

SQL commands below:

- create a table with 3 minutes TTL and UUID primary key (does this matter? I noticed the logic tests in `sql/logictest/testdata/logic_test/row_level_ttl` appeared to use a table with `INT` primary key)
- add rows
- `SELECT *` after the TTL is supposed to have expired, and the rows are not deleted

(Am I just holding it wrong?)

```
root@127.0.0.1:26257/movr> create database test;
create database test;
CREATE DATABASE


Time: 9ms total (execution 9ms / network 0ms)

root@127.0.0.1:26257/movr> use test;
use test;
SET


Time: 6ms total (execution 6ms / network 0ms)

root@127.0.0.1:26257/test> CREATE TABLE events (
  id UUID PRIMARY KEY default gen_random_uuid(),
  description TEXT,
  inserted_at TIMESTAMP default current_timestamp()
) WITH (ttl_expire_after = '3 minutes');

Time: 17ms total (execution 16ms / network 0ms)

root@127.0.0.1:26257/test> select current_timestamp();
select current_timestamp();
        current_timestamp
---------------------------------
  2022-03-15 15:36:43.530233+00
(1 row)


Time: 0ms total (execution 0ms / network 0ms)

root@127.0.0.1:26257/test> SHOW CREATE TABLE events;
SHOW CREATE TABLE events;
  table_name |                                                                                           create_statement
-------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  events     | CREATE TABLE public.events (
             |     id UUID NOT NULL DEFAULT gen_random_uuid(),
             |     description STRING NULL,
             |     inserted_at TIMESTAMP NULL DEFAULT current_timestamp():::TIMESTAMP,
             |     crdb_internal_expiration TIMESTAMPTZ NOT VISIBLE NOT NULL DEFAULT current_timestamp():::TIMESTAMPTZ + '00:03:00':::INTERVAL ON UPDATE current_timestamp():::TIMESTAMPTZ + '00:03:00':::INTERVAL,
             |     CONSTRAINT events_pkey PRIMARY KEY (id ASC)
             | ) WITH (ttl = 'on', ttl_automatic_column = 'on', ttl_expire_after = '00:03:00':::INTERVAL)
(1 row)


Time: 25ms total (execution 25ms / network 0ms)

root@127.0.0.1:26257/test> INSERT INTO events (description) VALUES ('a thing', 'another thing', 'yet another thing');
INSERT INTO events (description) VALUES ('a thing', 'another thing', 'yet another thing');
ERROR: INSERT has more expressions than target columns, 3 expressions for 1 targets
SQLSTATE: 42601
root@127.0.0.1:26257/test> INSERT INTO events (description) VALUES ('a thing'), ('another thing'), ('yet another thing');
INSERT INTO events (description) VALUES ('a thing'), ('another thing'), ('yet another thing');
INSERT 3


Time: 2ms total (execution 2ms / network 0ms)

root@127.0.0.1:26257/test> select current_timestamp();
select current_timestamp();
        current_timestamp
---------------------------------
  2022-03-15 15:38:22.331469+00
(1 row)


Time: 0ms total (execution 0ms / network 0ms)

root@127.0.0.1:26257/test> select current_timestamp();
select current_timestamp();
        current_timestamp
---------------------------------
  2022-03-15 15:40:15.983517+00
(1 row)


Time: 1ms total (execution 0ms / network 0ms)

root@127.0.0.1:26257/test> select * from events;
select * from events;
                   id                  |    description    |        inserted_at
---------------------------------------+-------------------+-----------------------------
  48410b9e-9b30-4737-9e97-967b7d8d826b | yet another thing | 2022-03-15 15:37:54.786547
  a6924763-13f7-4d59-8a32-7aef6b52e12b | another thing     | 2022-03-15 15:37:54.786547
  cdb7ce47-2d20-4eb6-9405-31703e5de0bb | a thing           | 2022-03-15 15:37:54.786547
(3 rows)


Time: 2ms total (execution 2ms / network 0ms)

root@127.0.0.1:26257/test> select current_timestamp();
select current_timestamp();
        current_timestamp
---------------------------------
  2022-03-15 15:40:29.072825+00
(1 row)


Time: 1ms total (execution 0ms / network 0ms)

root@127.0.0.1:26257/test> select current_timestamp();
select current_timestamp();
        current_timestamp
---------------------------------
  2022-03-15 15:40:40.153177+00
(1 row)


Time: 0ms total (execution 0ms / network 0ms)

root@127.0.0.1:26257/test> select current_timestamp();
select current_timestamp();
        current_timestamp
---------------------------------
  2022-03-15 15:42:44.718892+00
(1 row)


Time: 1ms total (execution 0ms / network 0ms)

root@127.0.0.1:26257/test> select * from events;
select * from events;
                   id                  |    description    |        inserted_at
---------------------------------------+-------------------+-----------------------------
  48410b9e-9b30-4737-9e97-967b7d8d826b | yet another thing | 2022-03-15 15:37:54.786547
  a6924763-13f7-4d59-8a32-7aef6b52e12b | another thing     | 2022-03-15 15:37:54.786547
  cdb7ce47-2d20-4eb6-9405-31703e5de0bb | a thing           | 2022-03-15 15:37:54.786547
(3 rows)


Time: 15ms total (execution 2ms / network 13ms)
```

Jira issue: CRDB-13794

Update: checked the TTL job and it says it is hourly?  Sorry if this is noise, I don't understand this

```
SELECT * FROM [SHOW SCHEDULES] WHERE label LIKE 'row-level-ttl-%';
          id         |       label       | schedule_status |        next_run        | state | recurrence | jobsrunning | owner |            created            |     command
---------------------+-------------------+-----------------+------------------------+-------+------------+-------------+-------+-------------------------------+-------------------
  744777962902716417 | row-level-ttl-114 | ACTIVE          | 2022-03-15 16:00:00+00 | NULL  | @hourly    |           0 | root  | 2022-03-15 15:36:36.678064+00 | {"tableId": 114}
(1 row)
```

**Comment by rafiss** [comment]
To make the TTL job run more frequently, you need the `ttl_job_cron` setting. More details in https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/20220120_row_level_ttl.md

The `ttl_expire_after` setting says when the row is eligible for deletion, not when it actually gets deleted.

The behavior if expired rows still appearing in output is discussed at https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/20220120_row_level_ttl.md#filtering-out-expired-rows

**Comment by otan** [comment]
turns out i indeed was able to run into it.

**Comment by otan** [comment]
It is a blocker due to it affecting anything with TTL

**Comment by otan** [comment]
Oh oops I see
